_posts/2020-02-13-spring-boot-application-events-explained.md

```diff
@@ -0,0 +1,263 @@
+
+Now that we have a `UserCreatedEvent` by extending `ApplicationEvent` and `UserRemovedEvent` without extending. When the object is not an `ApplicationEvent`, Spring will wrap it in a `PayloadApplicationEvent` for us. In order to publish it to its respective listeners, we need an instance of`ApplicationEventPublisher`:
+
+```java
+@Component
+class Publisher {
+	@Autowired
+	private ApplicationEventPublisher publisher;
+
+	void publishEvent(final String name) {
+		// Publishing event created by extending ApplicationEvent
+		publisher.publishEvent(new UserCreatedEvent(this, name));
+		// Publishing a object as an event
+		publisher.publishEvent(new UserRemovedEvent(name));
+	}
+}
+```
+
+## Listening to an Application Event
+
+Now that we know how to create and publish a custom event, let's see how we can listen to the event. An event can have multiple listeners doing different work based on the application requirements.
+
+There are two ways to define a listener. We can either use the `@EventListener` annotation or implement the `ApplicationListener` interface. In either case, the listener class has to be managed by Spring.
+
+### Annotation-Driven
+
+Starting with Spring 4.1 it's now possible to simply annotate a method of a managed bean with `@EventListener` to automatically register an `ApplicationListener` matching the signature of the method. No additional configuration is necessary with annotation-driven configuration enabled. If your method should listen to several events or if you want to define it with no parameter at all, the event types can also be specified on the annotation itself. Example: `@EventListener({ContextStartedEvent.class, ContextRefreshedEvent.class})`.
+For the methods annotated with `@EventListener` and defined as a non-void return type, Spring will send that result as a new event for us.
+
+```java
+@Component
+class UserRemovedListener {
+
+	@EventListener
+	ReturnedEvent handleUserRemovedEvent(UserRemovedEvent event) {
+		System.out.println(String.format("User removed (@EventListerner): %s", event.getName()));
+		// Spring will send ReturnedEvent as a new event
+		return new ReturnedEvent();
+	}
+
+	// Listener to receive the event returned by Spring
+	@EventListener
+	void handleReturnedEvent(ReturnedEvent event) {
+		System.out.println("Returned Event Called");
+	}
+	...
+}
+```
+
+Spring allows our listener to be triggered only in certain circumstances if we specify a `condition` by defining a boolean SpEL expression. The event will only be handled if the expression evaluates to `true` or one of the following strings: "true", "on", "yes", or "1". Method arguments are exposed via their names. The condition expression also exposes a “root” variable referring to the raw `ApplicationEvent` (`#root.event`) and the actual method arguments `(#root.args)`
```

Comment:
Again, explaining conditions before the reader had a chance to see the code example. Please re-structure.